### PR TITLE
Refactor certbot snap wrapper

### LINKED ISSUE #8280: jq error in Certbot snap
On a clean Debian 10 VPS from DigitalOcean I ran:
```
apt update
apt install snapd -y
snap install --classic certbot
/snap/bin/certbot --version
```
The last command fails with the output:
```
jq: error (at <stdin>:0): Cannot iterate over null (null)
```
and the exit status 5.

I'm not sure how widespread the problem is here or what the root cause is, but I wasn't able to reproduce it on my Ubuntu 20.04 machine.

cc @adferrand, @alexzorin

EDIT: After the discussion below, the changes we decided to make here are:

1. ~~Help users avoid installing the Certbot snap with an old version of snapd. Probably the best way to do this it update our instructions on https://certbot.eff.org to have people manually update it, but I personally think it's worth a little time to see if there's an easy way for us to automate this (outside of trying to run `snap install core` ourselves in our snap which I don't think we should do).~~ Done in https://github.com/certbot/website/pull/650.
2. ~~Improve the error message for people who somehow get themselves into this situation anyway. I still think it should be fatal but we can certainly make the description of the problem and how to fix it more clear. As part of this, I think we may want to consider rewriting `certbot.wrapper` in Python as suggested by Alex at https://github.com/certbot/certbot/issues/8280#issuecomment-690768231, but I personally think we can leave whether or not this is done up to whoever implements this.~~ Done in https://github.com/certbot/certbot/pull/8313.
3. ~~Report a bug against https://github.com/canonical-web-and-design/snapcraft.io asking them to fix the Debian instructions to avoid this problem.~~ Done in https://github.com/canonical-web-and-design/snapcraft.io/pull/3074.

### PR DESCRIPTION
**Pull request by adferrand**
Partial fix for #8280

This PR refactors the bash script wrapper for snap (`/certbot.wrapper`) into certbot python codebase. Here are the keypoints of this refactoring:
* the wrapping is applied when `main` function from `certbot._internal.main` is called if environment variable `CERTBOT_SNAPPED` is `True`, which is set during the snap build
* the initial bash script wrapper  is removed, simplifying `snap/snapcraft.yaml` by removing the `certbot.wrapper` part
* the dependency to `curl` and `jq` binaries are removed
* the failure during requesting the snapd socket is correctly handled, and displays an informative message in order to correct the situation, as required by #8280

One side note about the modifications done to `app.certbot.command` in `snapcraft.yaml`. Normally calling `bin/certbot` should be sufficient and it is effectively under a normal situation (`core` snap up-to-date). However in the same situation than when the problem occurs in #8280, using `bin/certbot` makes the snap raise an exception about `certbot.main` module that cannot be found.

It seems that when `core` snap is not up-to-date (in Debian for instance with default `snapd` installation), the shebang `/usr/bin/env python3` in the `bin/certbot` wrapper is wrongly resolved to the host Python, instead of the snap Python. It is working as expected if `core` snap is up-to-date. One way to fix that is to keep a bash script wrapper, because in this case, it is the `PATH` value that matters to resolve the Python interpreter, and `PATH` is correctly set up to resolve it from the snap first.

However to keep the simplification provided by the wrapper removal, I prefered to use `bin/python3 $SNAP/bin/certbot` as `command` to explicitly target the correct Python interpreter. Again normally it is not needed because everything is working correctly with a `core` snap up-to-date, but since the root purpose of all of this is to target bad situations, well, it is better to have a snap that is effectively able to start to display the informative message...

### CODE REVIEW

File: certbot/certbot/_internal/snap_config.py
Comment (atombrella) [inline certbot/certbot/_internal/snap_config.py:18]: How about using a dictionary for this instead?

Comment (adferrand) [reply certbot/certbot/_internal/snap_config.py:24]: Totally! I was still designing this from the bash script...

File: snap/snapcraft.yaml
Comment (alexzorin) [inline snap/snapcraft.yaml:23]: I wasn't able to build this with `snapcraft --use-lxd`:

    ....
    Staging certbot
    Staging shared-metadata
    Priming certbot
    Priming shared-metadata
    Failed to generate snap metadata: The specified command '$SNAP/bin/python3 $SNAP/bin/certbot' defined in the app 'certbot' is not executable.
    Run the same command again with --debug to shell into the environment if you wish to introspect this failure.

Comment (adferrand) [reply snap/snapcraft.yaml:23]: I may have been to much zealous by adding `$SNAP`. I am retrying, maybe I did not cleaned properly before rebuilding.

Comment (adferrand) [reply snap/snapcraft.yaml]: Ok, super weird. The executable in the command (`bin/python3`) must be relative to the snap installation path, but is not executed with the current working directory set to it, because the script argument (`bin/certbot`) is not found if also set relatively. It must be absolute. I suppose snapd just appends in some way the path to the executable dynamically.

Finally it is working with `bin/python3 $SNAP/bin/certbot`.

**Comment by adferrand** [comment]
I tested the snap on Ubuntu 20.04 (locally with WSL2) + Debian 9 & 10 on DigitalOcean and noticed the expected behavior: error with informative message with ancient `core`, correct execution of Certbot otherwise.

This deserve a run of snap integration tests: https://dev.azure.com/certbot/certbot/_build/results?buildId=2688&view=results

File: certbot/certbot/_internal/snap_config.py
Comment (alexzorin) [inline certbot/certbot/_internal/snap_config.py:56]: A small edge case:

One of the scenarios that Brad encountered was an Ubuntu AMI where `core` was installed at a very old version. snapd autorefresh had not yet had a chance to update it, because the server is freshly created.

AFAIK, `snap install` does not imply `snap refresh`. To close the loop here, we could add `; sudo snap refresh core`.

There's also the choice to do nothing, because the output from `snap install core` leaves a signpost for the user:

    snap "core" is already installed, see 'snap help refresh'

File: certbot/certbot/_internal/main.py
Comment (alexzorin) [inline certbot/certbot/_internal/main.py:1328]: Do you think this would benefit from additional error handling, if something does go wrong in `prepare_env`? 

Say, if `snapd` is stopped or isn't ready and returns a 5xx, [this is what we get](https://gist.github.com/alexzorin/ba2e98cabf658f0463a8cafb73115d5b). IMO it would be good to only print something like:

    An error occurred while fetching Certbot snap plugins: requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionRefusedError(111, 'Connection refused'))

but it does run the risk of obscuring too much without having a logging system initialized yet.

Comment (adferrand) [reply certbot/certbot/_internal/main.py:1328]: Well, the error output that would emerge from the wrapper phase will be the first thing a regular user who just installed Certbot through snap would see. This stacktrace is hideous, I think we need to do something about it.

Indeed having a short message describing that something wrong happened, with headings, would be nice.

Given what the wrapper does, I think we can move the wrapper call after `log.pre_arg_parse_setup()` without side effects. This way, we can benefit from the minimal logging system set up at that time to handle a little better any error occurring during the initialization phase.

What do you think of this?

Comment (alexzorin) [reply certbot/certbot/_internal/main.py:1328]: I would personally prefer to be cautious; keep it as the first call and put a broad exception handler around it. But this is just paranoia, I don't see anything specifically wrong with waiting for the logging system.

Comment (adferrand) [reply certbot/certbot/_internal/main.py:1330]: The setup is about setting some environment variables and a cli arg. None of this will impact the logging system, I think we are safe.

I move the logic after `log.pre_arg_parse_setup()`. This way we get a "nice" output and a log stored in the temporary directory when an unexpected error occurs.

File: certbot/certbot/_internal/snap_config.py
Comment (adferrand) [reply certbot/certbot/_internal/snap_config.py]: I added your suggestion. It avoids one step for the user to solve the situation.

**Comment by bmw** [comment]
Thanks a lot for getting started on the review here @alexzorin!

For the sake of sharing work of reviewing PRs, do you think you're able and interested in being the primary reviewer here or should I also review this? I could also just double check specific pieces of this if you'd like.

File: certbot/certbot/_internal/main.py
Comment (alexzorin) [reply certbot/certbot/_internal/main.py:1330]: That's way better. 

1. Since we have a logger and except hook available now, shall we use it in `snap_config.py` instead of `sys.stderr.write` and `sys.exit`?

2. I think we should keep improving the below error output, so that it is clear to the user that it is `snapd`-related and not actually a network issue. Maybe if something in `requests` errors, error log an explanation (`Error occurred fetching .... make sure the snapd service is running`), then re-raise.
```
$ certbot --version
An unexpected error occurred:
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionRefusedError(111, 'Connection refused'))
Please see the logfile '/tmp/tmpiicnb1f3/log' for more details.
```

Comment (adferrand) [reply certbot/certbot/_internal/main.py:1330]: Done.

File: snap/snapcraft.yaml
Comment (alexzorin) [inline snap/snapcraft.yaml:23]: Unfortunately, Debian 9 doesn't like the `$`:

    root@greek-totem:~# snap install --dangerous certbot_1.8.0-33-gf659124ed_amd64.snap
    error: cannot read snap file: app description field 'command' contains illegal "bin/python3 $SNAP/bin/certbot" (legal: '^[A-Za-z0-9/. _#:-]*$')

It looks like it only became an option in `2.29`: https://github.com/snapcore/snapd/pull/3995.

Is there any hope at all of using a relative path here?

Comment (adferrand) [reply snap/snapcraft.yaml:23]: Oh. No without an up-to-date snapd it was not working... Given the limitations of what can be used in the command, passing a python command like `bin/python -c "from certbot.main import main; main()"`.

The only remaining option I can think of now is a package call like `bin/python - m certbot`, but I think `__main__.py` is needed in this case.

Comment (adferrand) [reply snap/snapcraft.yaml]: I did that. It works locally and on a Debian 10 DigitalOcean droplet. I could not test the Debian 9 one, because without `snap install core`, the entire `snap` command is not functional as it is unable to download any base snap.

Comment (alexzorin) [reply snap/snapcraft.yaml]: True, I suppose this can only be fixed by documentation for Debian 9.

File: certbot/certbot/_internal/snap_config.py
Comment (alexzorin) [inline certbot/certbot/_internal/snap_config.py:46]: Any reason not to use the logger here?

Comment (adferrand) [reply certbot/certbot/_internal/snap_config.py:46]: No reason at all. I update that.

Comment (adferrand) [reply certbot/certbot/_internal/snap_config.py]: More precisely I used `certbot.errors.Error` here, that will be properly handled by the logging system.

File: snap/snapcraft.yaml
Comment (alexzorin) [inline snap/snapcraft.yaml:23]: Something is going wrong with logging and error handling because of this change:

    $ sudo -E ./venv3/bin/certbot --apache
    Saving debug log to /var/log/letsencrypt/letsencrypt.log
    The apache plugin is not working; there may be problems with your existing configuration.
    The error was: NoInstallationError('Cannot find Apache executable apache2ctl')

  versus what happens in the snap right now:

    $ sudo -E ./venv3/bin/python3 -m certbot --apache
    Saving debug log to /var/log/letsencrypt/letsencrypt.log

If I understand right, you made this change because of Debian 9's dislike of `$SNAP`.

However, as you correctly pointed out, we can't really do anything about Debian 9 anyway, shall we revert this to what it previously was?

Comment (adferrand) [reply snap/snapcraft.yaml]: Ok. I reverted to the `bin/python3 $SNAP/bin/certbot` approach. I also double-checked that, on top of having the proper error handling on snap connection failing, the `/snap/bin/certbot --apache` call is failing also with the proper error.

**Comment by adferrand** [comment]
Test focusing on snaps from the full test pipeline is here: https://dev.azure.com/certbot/certbot/_build/results?buildId=2756&view=results

File: snap/snapcraft.yaml
Comment (alexzorin) [inline snap/snapcraft.yaml:23]: What does `-m certbot` do here?

Comment (adferrand) [reply snap/snapcraft.yaml]: My bad, its fixed.

**Comment by adferrand** [comment]
Updated tests focusing on snap here: https://dev.azure.com/certbot/certbot/_build/results?buildId=2758&view=results

**Review by alexzorin — APPROVED** [review]
LGTM, in combination with website changes.

**Comment by adferrand** [comment]
@alexzorin I update the PR to keep instructions sync with https://github.com/certbot/website/pull/650. As a consequence GitHub dismissed your review.

**Review by alexzorin — APPROVED** [review]

**Comment by bmw** [comment]
Thanks a lot for working on this you two!

Is there something that either of you think should be checked by yourselves or someone else before merging this PR or is it ready to go?

**Comment by alexzorin** [comment]
All good from me.

**Pull request merged**
